app/javascript/packages/pwned-passwords/pairing-heap.js

```diff
@@ -0,0 +1,142 @@
+/**
+ * @template V
+ * @typedef {(a: V, b: V) => number} Comparator
+ */
+
+/**
+ * @template V
+ * @typedef {{ value: V, head?: Heap<V>, next?: Heap<V> }} Heap
+ */
+
+/**
+ * @template V
+ */
+class PairingHeap {
+  /** @type {Comparator<V>} */
+  comparator;
+
+  /** @type {Heap<V> | undefined} */
+  heap;
+
+  /** @type {number} */
+  length = 0;
+
+  /**
+   * @param {Comparator<V>} comparator
+   */
+  constructor(comparator) {
+    this.comparator = comparator;
+  }
+
+  *[Symbol.iterator]() {
+    if (!this.heap) {
+      return;
+    }
+
+    /** @type {Heap<V>[]} */
+    const queue = [this.heap];
+    for (let i = 0; i < queue.length; i++) {
+      const { value, head, next } = queue[i];
+
+      yield value;
+
+      if (next) {
+        queue.push(next);
+      }
+
+      if (head) {
+        queue.push(head);
+      }
+    }
+  }
+
+  /**
+   * Returns the top value of the heap.
+   *
+   * @return {V}
+   */
+  peek() {
+    return /** @type {Heap<V>} */ (this.heap).value;
+  }
+
+  /**
+   * Adds a new value to the heap.
+   *
+   * @param {V} value
+   */
+  push(value) {
+    if (this.heap) {
+      this.heap = this.#merge(this.heap, { value });
+    } else {
+      this.heap = { value };
+    }
+
+    this.length++;
+  }
+
+  /**
+   * Removes and returns the top value from the heap.
+   *
+   * @return {V}
+   */
+  pop() {
+    const { value, head } = /** @type {Heap<V>} */ (this.heap);
+
+    this.heap = this.#mergePairs(head);
+    this.length--;
+
+    return value;
+  }
+
+  /**
+   * Melds pairs of heaps to a new heap.
+   *
+   * @param {Heap<V> | undefined} heap
+   * @return {Heap<V> | undefined}
+   */
+  #mergePairs(heap) {
+    if (heap === undefined) {
+      return;
+    }
+
+    const { next } = heap;
+    if (!next) {
+      return heap;
+    }
+
+    const nextSequence = next.next;
+    heap.next = undefined;
+    next.next = undefined;
+
+    return this.#merge(this.#merge(heap, next), this.#mergePairs(nextSequence));
```

Comment:
When trying to download the full range at maximum size, I hit a "Maximum call stack size exceeded" error here 😬 

I guess it's not too unexpected that this could happen, since the tree could become quite large and we're doing recursion here.

Might need to see if I can rewrite this as non-recursive, or reevaluate the approach with managing the list.

Reply:
Ok, I think I was successfully able to rewrite this to avoid recursion in d65fb66. At least, I was able to run the full download longer than I was previously.

Per "reevaluate the approach" and hitting some roadbumps, I was curious how a naive approach using an array might fare as an alternative to this data structure. 

Benchmarking it, the array solution does pretty well when the maximum size is small (e.g. 1,000), but it starts to slow to a crawl at the size we're looking for (i.e. 3,000,000).

```js
import Benchmark from "benchmark";
import PairingHeap from './pairing-heap.js';

var suite = new Benchmark.Suite();

const comparator = (a, b) => a - b;
const comparatorReverse = (a, b) => b - a;
const heap = new PairingHeap(comparator);
const array = [];
const MAX_SIZE = 3_000_000;

suite
  .add("PairingHeap", function () {
    const value = Math.random();
    if (heap.length >= MAX_SIZE) {
      if (comparator(heap.peek(), value) < 0) {
        heap.pop();
      } else {
        return;
      }
    }

    heap.push(value);
  })
  .add("Array", function () {
    const value = Math.random();
    if (array.length >= MAX_SIZE) {
      if (comparator(array[array.length - 1], value) < 0) {
        array[array.length - 1] = value;
      } else {
        return;
      }
    } else {
      array.push(value);
    }

    array.sort(comparatorReverse);
  })
  .on("cycle", function (event) {
    console.log(String(event.target));
  })
  .on("complete", function () {
    console.log("Fastest is " + this.filter("fastest").map("name"));
  })
  .run({ async: true });
```

With `MAX_SIZE = 1_000`:

```
PairingHeap x 96,075,173 ops/sec ±0.32% (93 runs sampled)
Array x 146,766,288 ops/sec ±0.27% (97 runs sampled)
Fastest is Array
```

With `MAX_SIZE = 3_000_000`:

```
PairingHeap x 714,216 ops/sec ±87.98% (9 runs sampled)
Array x 1,051 ops/sec ±84.71% (5 runs sampled)
Fastest is PairingHeap
```